gapis/service/types/types.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/google/gapid/core/data/pod"
+	"github.com/google/gapid/gapis/memory"
+)
+
+type typeData struct {
+	tp *Type
+	rt reflect.Type
+}
+
+var type_map = make(map[uint64]*typeData)
+
+func AddType(i uint64, t *Type, rt reflect.Type) {
+	if _, ok := type_map[i]; ok {
+		panic(fmt.Errorf("Error identical types exist %+v", i))
+	}
+	type_map[i] = &typeData{t, rt}
+}
+
+func GetType(i uint64) (*Type, error) {
+	t, ok := type_map[i]
+	if !ok {
+		return nil, fmt.Errorf("Could not find type %v", i)
+	}
+	return t.tp, nil
+}
+
+func GetReflectedType(i uint64) (reflect.Type, error) {
+	t, ok := type_map[i]
+	if !ok {
+		return nil, fmt.Errorf("Could not find type %v", i)
+	}
+	return t.rt, nil
+}
+
+// GetTypeIndex returns the index of the type. This must be
+// kept in sync with api_types.go.tmpl
```

Comment:
Isn't there a less error-prone way to keep both files in sync?

Reply:
I was hoping, we could use some globals, that may  help. Let me see what I can do.